pythia/utils/text_utils.py

```diff
@@ -289,3 +289,91 @@ def best_score(self):
             i = self.complete_seqs_scores.index(max(self.complete_seqs_scores))
             captions = torch.FloatTensor(self.complete_seqs[i]).unsqueeze(0)
         return captions
+
+
+class NucleusSampling:
+    def __init__(self, vocab, threshold=0.8):
+        self.vocab = vocab
+        self.vocab_size = vocab.get_size()
+        # Threshold for sum of probability
+        self.threshold = threshold
+        # Lists to store completed sequence
+        self.complete_seq = []
+
+    def init_batch(self, sample_list):
+        setattr(
+            self,
+            "seq",
+            sample_list.answers.new_full(
+                (1, 1), self.vocab.SOS_INDEX, dtype=torch.long
+            ),
+        )
+        setattr(
+            self,
+            "top_m_scores",
+            sample_list.answers.new_zeros((1, 1), dtype=torch.float),
+        )
+        # Add a dim and duplicate the tensor beam_size times across that dim
+        print("before: ", sample_list.image_feature_0)
+        sample_list.image_feature_0 = (
+            sample_list.image_feature_0.unsqueeze(1).expand(-1, 1, -1, -1).squeeze(0)
+        )
+        print("after: ", sample_list.image_feature_0)
+        return sample_list
+
+    def sample(self, t, data, scores):
+        # Convert scores to probabilities
+        scores = torch.nn.functional.softmax(scores, dim=1)
+        # Sort scores in descending order and then select the top m elements having sum more than threshold.
+        # We get the top_m_scores and their indices top_m_words
+        if t == 0:
+            self.top_m_scores, top_m_words = scores[0].sort(0, True)
+        else:
+            self.top_m_scores, top_m_words = scores.view(-1).sort(0, True)
+
+        lastIndex = 0
+        scoreSum = 0
+        for score in self.top_m_scores:
+            lastIndex += 1
+            scoreSum += score
+            if scoreSum >= self.threshold:
+                break
+
+        self.top_m_scores = torch.div(self.top_m_scores[:lastIndex], scoreSum)
+        top_m_words = top_m_words[:lastIndex]
+        # Zero value inside prev_word_inds because we are predicting a single stream of output.
+        prev_word_ind = torch.tensor([0])
+        # Get next word based on probabilities of top m words.
+        next_word_ind = top_m_words[torch.multinomial(self.top_m_scores, 1)]
+        # Add next word to sequence
+        self.seq = torch.cat(
+            [self.seq[prev_word_ind], next_word_ind.unsqueeze(1)], dim=1
+        )
+        # Check if sequence is complete
```

Comment:
Redundant code in beam_search and here. Consider making this a function and calling it here.

Reply:
Agreed. Should I make that function in utils/general.py?